### ISSUE TITLE
can't connect to wss://URL caused by self-sign certificate

### ISSUE THREAD

**Issue by ozzzzchen** [mqttjs/MQTT.js#288]
<h3>Environment:</h3>
**MQTT Broker:**
1. Windows Server 2012 Datacenter R2 on Microsoft Azure VM
2. Node.js v0.10.38
3. iisnode v0.2.18
4. Mosca v0.29.0
5. IIS opened https with port 443 and created a self-sign certificate
**MQTT Client:**
1. Windows 7 Ultimate
2. Node.js v0.10.38
3. MQTT.js v1.2.0

<h3>Test on HiveMQ WebSocket Client:</h3>
1. [HiveMQ WebSocket Client](http://www.hivemq.com/demos/websocket-client/)
2. first connect to [https://URL](https://waoz.cloudapp.net/) and trust the certificate from server on browser
3. successfully connect to server with parameter:
```
Host: waoz.cloudapp.net/Mosca/
Port: 443
SSL: checked
```

<h3>Test on MQTT.js:</h3>
1. connect to ws://URL/Mosca/ is work fine
2. connect to wss://URL:443/Mosca/ is failed
3. here is codes:
```javascript
var mqtt = require('mqtt');

//var client = mqtt.connect('ws://waoz.cloudapp.net/Mosca/');
var client = mqtt.connect('wss://waoz.cloudapp.net:443/Mosca/');

client.on('connect', function ()
{
    console.log('connect');
    client.end();
});

client.on('offline', function ()
{
    console.log('offline');
});

client.on('close', function ()
{
    console.log('close');
    client.end();
});
```
output is:
```
offline
close
```

<h3>Questions:</h3>
1. How do I successfully connect to wss://URL:443/Mosca/ with MQTT.js ?
2. Should I trust the certificate from server like browser done and how ?
3. Do I need to create a client certificate?

many thanks!

**Comment by mcollina** [comment]
1. exactly like you did, but you either need to use a valid certificate, or tell MQTT.js what to do. The default is to be *secure*, so self-signed certificate are not allowed by default.
2. There are various options, look at https://nodejs.org/api/tls.html#tls_tls_connect_options_callback for everything, as the options are passed as-is to `tls.connect`. The simple one is to pass `rejectUnauthorized: false` in the options.
3. No.

**Comment by ozboy** [comment]
Thank you for your advice.
I have passed the option directly as the following but it was not work.
```javascript
var client = mqtt.connect('wss://waoz.cloudapp.net:443/Mosca/', {rejectUnauthorized:false});
```


I found the options didn't pass to websocket library in function *buildBuilder* which is in /lib/connect/ws.js
```javascript
function buildBuilder (client, opts) {
  var host = opts.hostname || 'localhost',
    port = String(opts.port || 80),
    path = opts.path || '/',
    url = opts.protocol + '://' + host + ':' + port + path,
    ws = websocket(url, {
      protocol: 'mqttv3.1'
    });

  return ws;
}
```


It is works now after some modification:
```javascript
function buildBuilder (client, opts) {
    var wsOpt = {};
    wsOpt['protocol'] = 'mqttv3.1';
    if(opts.protocol == 'wss')
        if(opts.hasOwnProperty('rejectUnauthorized'))
            wsOpt['rejectUnauthorized'] = opts.rejectUnauthorized;
    var host = opts.hostname || 'localhost',
        port = String(opts.port || 80),
        path = opts.path || '/',
        url = opts.protocol + '://' + host + ':' + port + path,
        ws = websocket(url, wsOpt);

    return ws;
}
```

**Comment by mcollina** [comment]
Can you send a PR for these fixes?

**Comment by PritamUpadhyay** [comment]
@ozzzzchen ; Did you find any way to connect?

**Comment by PritamUpadhyay** [comment]
@mcollina :Any update for this issue?

**Comment by mcollina** [comment]
If somebody sends a PR with tests, I'll be happy to review, test and merge :).

**Comment by ozzzzchen** [comment]
The issue has already solved in PR https://github.com/mqttjs/MQTT.js/pull/289

**Comment by ozzzzchen** [comment]
The issue has already solved in PR #289

**Comment by renerlemes** [comment]
How to use CA signed server certificate?
